INFO and ERROR logging should come out in order by default

Right now, info-level logging output is not line-buffered by default, whereas error-level logging output is. This is a problem for usability, since it causes extremely confusing logs when an error message in one process is flushed before a much later info message in a simultaneous process. The buffering of stdout, while appropriate for some Linux programs, is inappropriate when diagnostic logging output is emitted from both streams stderr and stdout.

The consequences of having `RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED=1` when it should be 0 is a tiny (if noticeable) performance hit. The consequences of having it 0 when it should be 1 is incomprehensibly misordered logging output, often with info messages not emitted until a node dies, and no sequence numbers nor timestamps to indicate this misordering. Because of this, logging should default `RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED` to 1 if unset.

An example of bad logging in a complex, multi-node launch. It's very hard to determine the order things happen in between multiple nodes. If line flushing, it would be clear that `dwb_controller` is created *before* `amcl` terminates.

```
ros2 launch openrover_demo nav2.launch.py
[INFO] [launch]: All log files can be found below /home/snuc/.ros/log/2019-06-10-17-13-22-508548-rhea-30016
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [lifecycle_manager-1]: process started with pid [30046]
[INFO] [map_server-2]: process started with pid [30047]
[INFO] [amcl-3]: process started with pid [30048]
[INFO] [world_model-4]: process started with pid [30052]
[INFO] [navfn_planner-5]: process started with pid [30054]
[INFO] [motion_primitives_node-6]: process started with pid [30063]
[INFO] [bt_navigator-7]: process started with pid [30071]
[INFO] [dwb_controller-8]: process started with pid [30078]
[lifecycle_manager-1] [WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
[amcl-3] [WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
[navfn_planner-5] [WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
[world_model-4] [WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
[dwb_controller-8] [WARN] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
^C[WARNING] [launch]: user interrupted with ctrl-c (SIGINT)
[amcl-3] [ERROR] [rclcpp]: failed to create wait set: the given context is not valid, either rcl_init() was not called or rcl_shutdown() was called., at /opt/ros/master/src/ros2/rcl/rcl/src/rcl/wait.c:130
[amcl-3] terminate called after throwing an instance of 'std::runtime_error'
[amcl-3]   what():  Failed to create wait set in Executor constructor
[INFO] [lifecycle_manager-1]: process has finished cleanly [pid 30046]
[ERROR] [amcl-3]: process has died [pid 30048, exit code -6, cmd '/home/snuc/ros2_ws/install/lib/nav2_amcl/amcl __node:=amcl __params:=/home/snuc/ros2_ws/install/share/openrover_demo/config/nav2.yaml'].
[INFO] [motion_primitives_node-6]: process has finished cleanly [pid 30063]
[INFO] [bt_navigator-7]: process has finished cleanly [pid 30071]
[lifecycle_manager-1] [INFO] [lifecycle_manager]: Creating
[lifecycle_manager-1] [INFO] [rclcpp]: signal_handler(signal_value=2)
[lifecycle_manager-1] [INFO] [lifecycle_manager]: Destroying
[motion_primitives_node-6] [INFO] [motion_primitives]: Configuring Spin
[motion_primitives_node-6] [INFO] [motion_primitives]: Robot: Creating
[motion_primitives_node-6] [INFO] [motion_primitives]: Configuring BackUp
[motion_primitives_node-6] [INFO] [motion_primitives]: Robot: Creating
[motion_primitives_node-6] [INFO] [rclcpp]: signal_handler(signal_value=2)
[motion_primitives_node-6] [INFO] [motion_primitives]: Robot: Destroying
[motion_primitives_node-6] [INFO] [motion_primitives]: Robot: Destroying
[bt_navigator-7] [INFO] [bt_navigator]: Creating
[bt_navigator-7] [INFO] [rclcpp]: signal_handler(signal_value=2)
[bt_navigator-7] [INFO] [bt_navigator]: Destroying
[navfn_planner-5] [INFO] [navfn_planner]: Creating
[navfn_planner-5] [INFO] [rclcpp]: signal_handler(signal_value=2)
[navfn_planner-5] [INFO] [navfn_planner]: Destroying
[INFO] [navfn_planner-5]: process has finished cleanly [pid 30054]
[map_server-2] [INFO] [map_server]: Creating
[map_server-2] [INFO] [rclcpp]: signal_handler(signal_value=2)
[map_server-2] [INFO] [map_server]: Destroying
[world_model-4] [INFO] [world_model]: Creating
[world_model-4] [INFO] [global_costmap.world_model]: Creating
[world_model-4] [INFO] [rclcpp]: signal_handler(signal_value=2)
[world_model-4] [INFO] [world_model]: Destroying
[world_model-4] [INFO] [global_costmap.world_model]: Destroying
[INFO] [map_server-2]: process has finished cleanly [pid 30047]
[INFO] [world_model-4]: process has finished cleanly [pid 30052]
[dwb_controller-8] [INFO] [dwb_controller]: Creating
[dwb_controller-8] [INFO] [local_costmap.dwb_controller]: Creating
[dwb_controller-8] [INFO] [rclcpp]: signal_handler(signal_value=2)
[dwb_controller-8] [INFO] [dwb_controller]: Destroying
[dwb_controller-8] [INFO] [local_costmap.dwb_controller]: Destroying
[INFO] [dwb_controller-8]: process has finished cleanly [pid 30078]
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -45,6 +45,7 @@ set(rcutils_sources
   src/snprintf.c
   src/split.c
   src/strdup.c
+  src/strerror.c
   src/string_array.c
   src/string_map.c
   src/time.c

--- a/include/rcutils/strerror.h
+++ b/include/rcutils/strerror.h
@@ -37,7 +37,7 @@ extern "C"
  */
 RCUTILS_PUBLIC
 void
-rcutils_safe_strerror(char * buffer, size_t buffer_length);
+rcutils_strerror(char * buffer, size_t buffer_length);
 
 #ifdef __cplusplus
 }

--- a/include/rcutils/strerror.h
+++ b/include/rcutils/strerror.h
@@ -1,0 +1,33 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCUTILS__STRERROR_H_
+#define RCUTILS__STRERROR_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcutils/visibility_control.h"
+
+RCUTILS_PUBLIC
+void
+rcutils_safe_strerror(char * buffer, size_t buffer_length);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif  // RCUTILS__STRERROR_H_

--- a/include/rcutils/strerror.h
+++ b/include/rcutils/strerror.h
@@ -23,7 +23,8 @@ extern "C"
 #include "rcutils/visibility_control.h"
 
 /// Retrieve the string corresponding to the last system error.
-/** This function retrieves the value of errno, and calls the system-specific
+/**
+ * This function retrieves the value of errno, and calls the system-specific
  * equivalent of `strerror` on it, storing the output in the provided buffer.
  * If the error message is longer than the buffer, it will be truncated.
  * The memory for the c-string buffer that is passed in must be managed by the

--- a/include/rcutils/strerror.h
+++ b/include/rcutils/strerror.h
@@ -1,4 +1,4 @@
-// Copyright 2017 Open Source Robotics Foundation, Inc.
+// Copyright 2020 Open Source Robotics Foundation, Inc.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.
@@ -22,6 +22,18 @@ extern "C"
 
 #include "rcutils/visibility_control.h"
 
+/// Retrieve the string corresponding to the last system error.
+/* This function retrieves the value of errno, and calls the system-specific
+ * equivalent of `strerror` on it, storing the output in the provided buffer.
+ * If the error message is longer than the buffer, it will be truncated.
+ * The memory for the c-string buffer that is passed in must be managed by the
+ * caller.
+ *
+ * This function is thread-safe.
+ *
+ * \param[in/out] buffer the buffer in which to store the data
+ * \param[in] buffer_length the maximum length of the buffer
+ */
 RCUTILS_PUBLIC
 void
 rcutils_safe_strerror(char * buffer, size_t buffer_length);

--- a/include/rcutils/strerror.h
+++ b/include/rcutils/strerror.h
@@ -23,7 +23,7 @@ extern "C"
 #include "rcutils/visibility_control.h"
 
 /// Retrieve the string corresponding to the last system error.
-/* This function retrieves the value of errno, and calls the system-specific
+/** This function retrieves the value of errno, and calls the system-specific
  * equivalent of `strerror` on it, storing the output in the provided buffer.
  * If the error message is longer than the buffer, it will be truncated.
  * The memory for the c-string buffer that is passed in must be managed by the

--- a/src/logging.c
+++ b/src/logging.c
@@ -18,6 +18,7 @@ extern "C"
 #endif
 
 #include <ctype.h>
+#include <errno.h>
 #include <inttypes.h>
 #include <stdint.h>
 #include <stdio.h>
@@ -77,9 +78,6 @@ bool g_rcutils_logging_severities_map_valid = false;
 int g_rcutils_logging_default_logger_level = 0;
 
 static FILE * g_output_stream = NULL;
-static bool g_output_flush_stream = false;
-// A one-time print if we fail to flush
-static bool g_stream_flush_failure_reported = false;
 
 enum rcutils_colorized_output g_colorized_output = RCUTILS_COLORIZED_OUTPUT_AUTO;
 
@@ -88,11 +86,20 @@ rcutils_ret_t rcutils_logging_initialize(void)
   return rcutils_logging_initialize_with_allocator(rcutils_get_default_allocator());
 }
 
+enum rcutils_get_env_retval
+{
+  RCUTILS_GET_ENV_ERROR = -1,
+  RCUTILS_GET_ENV_ZERO = 0,
+  RCUTILS_GET_ENV_ONE = 1,
+  RCUTILS_GET_ENV_DEFAULT = 2,
+};
+
 // A utility function to get zero or one from an environment variable.  Returns
-// -1 if we failed to get the environment variable, 0 if the value in the
-// environment variable is something other than "0" or "1", 0 if the environment
-// variable is "0" or "", or "1" if the environment variable is "1".
-static int rcutils_get_env_var_zero_or_one(
+// RCUTILS_GET_ENV_ERROR if we failed to get the environment variable or if it was
+// something we don't understand.  Return RCUTILS_GET_ENV_ZERO if the value in the
+// environment variable is "0", RCUTILS_GET_ENV_ONE if the value in the environment
+// variable is "1", or RCUTILS_GET_ENV_DEFAULT if the environment variables is empty.
+static enum rcutils_get_env_retval rcutils_get_env_var_zero_or_one(
   const char * name, const char * zero_semantic,
   const char * one_semantic)
 {
@@ -102,24 +109,27 @@ static int rcutils_get_env_var_zero_or_one(
     fprintf(
       stderr, "Error getting environment variable "
       "%s: %s\n", name, ret_str);
-    return -1;
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("Error getting environment variable: %s", ret_str);
+    return RCUTILS_GET_ENV_ERROR;
   }
 
-  if (strcmp(env_var_value, "0") == 0 || strcmp(env_var_value, "") == 0) {
-    return 0;
+  if (strcmp(env_var_value, "") == 0) {
+    return RCUTILS_GET_ENV_DEFAULT;
   }
-
+  if (strcmp(env_var_value, "0") == 0) {
+    return RCUTILS_GET_ENV_ZERO;
+  }
   if (strcmp(env_var_value, "1") == 0) {
-    return 1;
+    return RCUTILS_GET_ENV_ONE;
   }
 
   fprintf(
     stderr,
     "Warning: unexpected value [%s] specified for %s. "
-    "Valid values are 0 (%s) or 1 (%s).  Default value 0 will be used.\n",
+    "Valid values are 0 (%s) or 1 (%s).",
     env_var_value, name, zero_semantic, one_semantic);
 
-  return 0;
+  return RCUTILS_GET_ENV_ERROR;
 }
 
 rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allocator)
@@ -139,37 +149,59 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     // are propagated immediately.  The user can choose to set the output stream
     // to stdout by setting the RCUTILS_CONSOLE_USE_STDOUT environment
     // variable to 1.
-    g_output_stream = stderr;
-    int retval = rcutils_get_env_var_zero_or_one(
+    enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
       "RCUTILS_CONSOLE_USE_STDOUT", "use stderr",
       "use stdout");
-    if (retval < 0) {
+    if (retval == RCUTILS_GET_ENV_ERROR) {
       return RCUTILS_RET_INVALID_ARGUMENT;
-    } else if (retval == 0) {
+    } else if (retval == RCUTILS_GET_ENV_DEFAULT || retval == RCUTILS_GET_ENV_ZERO) {
       g_output_stream = stderr;
-    } else if (retval == 1) {
+    } else if (retval == RCUTILS_GET_ENV_ONE) {
       g_output_stream = stdout;
     }
 
-    retval = rcutils_get_env_var_zero_or_one("RCUTILS_CONSOLE_FLUSH_STREAM", "no flush", "flush");
-    if (retval < 0) {
+    // Allow the user to choose how buffering on the stream works by setting
+    // RCUTILS_CONSOLE_BUFFERED_STREAM.  With an empty environment variable, use the
+    // default of the stream.  With a value of 0, force the stream to be unbuffered.
+    // With a value of 1, force the stream to be buffered.
+    retval =
+      rcutils_get_env_var_zero_or_one(
+      "RCUTILS_CONSOLE_BUFFERED_STREAM", "not buffered",
+      "buffered");
+    if (retval == RCUTILS_GET_ENV_ERROR) {
       return RCUTILS_RET_INVALID_ARGUMENT;
-    } else if (retval == 0) {
-      g_output_flush_stream = false;
-    } else if (retval == 1) {
-      g_output_flush_stream = true;
+    } else if (retval == RCUTILS_GET_ENV_ZERO) {
+      if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
+        fprintf(
+          stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Error setting stream buffering mode: %s", strerror(
+            errno));
+        return RCUTILS_RET_ERROR;
+      }
+    } else if (retval == RCUTILS_GET_ENV_ONE) {
+      if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
+        fprintf(
+          stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Error setting stream buffering mode: %s", strerror(
+            errno));
+        return RCUTILS_RET_ERROR;
+      }
     }
+    // Note that for retval == RCUTILS_GET_ENV_DEFAULT, we do nothing to leave it at default.
 
     retval = rcutils_get_env_var_zero_or_one(
       "RCUTILS_COLORIZED_OUTPUT", "force color",
       "force no color");
-    if (retval < 0) {
+    if (retval == RCUTILS_GET_ENV_ERROR) {
       return RCUTILS_RET_INVALID_ARGUMENT;
-    } else if (retval == 0) {
+    } else if (retval == RCUTILS_GET_ENV_ZERO) {
       g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE;
-    } else if (retval == 1) {
+    } else if (retval == RCUTILS_GET_ENV_ONE) {
       g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
     }
+    // Note that for retval == RCUTILS_GET_ENV_DEFAULT, we do nothing to leave it at default.
 
     // Check for the environment variable for custom output formatting
     const char * output_format;
@@ -892,16 +924,6 @@ void rcutils_logging_console_output_handler(
 
   if (RCUTILS_RET_OK == status) {
     fprintf(g_output_stream, "%s\n", output_array.buffer);
-
-    if (g_output_flush_stream) {
-      int flush_result = fflush(g_output_stream);
-      if (flush_result != 0 && !g_stream_flush_failure_reported) {
-        g_stream_flush_failure_reported = true;
-        fprintf(
-          stderr, "Error: failed to perform fflush on stdout, fflush return code is: %d\n",
-          flush_result);
-      }
-    }
   }
 
   // Only does something in windows

--- a/src/logging.c
+++ b/src/logging.c
@@ -92,14 +92,15 @@ enum rcutils_get_env_retval
   RCUTILS_GET_ENV_ERROR = -1,
   RCUTILS_GET_ENV_ZERO = 0,
   RCUTILS_GET_ENV_ONE = 1,
-  RCUTILS_GET_ENV_DEFAULT = 2,
+  RCUTILS_GET_ENV_EMPTY = 2,
 };
 
-// A utility function to get zero or one from an environment variable.  Returns
-// RCUTILS_GET_ENV_ERROR if we failed to get the environment variable or if it was
-// something we don't understand.  Return RCUTILS_GET_ENV_ZERO if the value in the
-// environment variable is "0", RCUTILS_GET_ENV_ONE if the value in the environment
-// variable is "1", or RCUTILS_GET_ENV_DEFAULT if the environment variables is empty.
+// A utility function to get zero or one from an environment variable.
+// Returns RCUTILS_GET_ENV_ERROR if we failed to get the environment variable
+// or if it was something we don't understand.
+// Return RCUTILS_GET_ENV_ZERO if the value in the environment variable is "0",
+// RCUTILS_GET_ENV_ONE if the value in the environment variable is "1", or
+// RCUTILS_GET_ENV_EMPTY if the environment variables is empty.
 static enum rcutils_get_env_retval rcutils_get_env_var_zero_or_one(
   const char * name, const char * zero_semantic,
   const char * one_semantic)
@@ -117,7 +118,7 @@ static enum rcutils_get_env_retval rcutils_get_env_var_zero_or_one(
   }
 
   if (strcmp(env_var_value, "") == 0) {
-    return RCUTILS_GET_ENV_DEFAULT;
+    return RCUTILS_GET_ENV_EMPTY;
   }
   if (strcmp(env_var_value, "0") == 0) {
     return RCUTILS_GET_ENV_ZERO;
@@ -155,7 +156,7 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
         fprintf(
           stderr,
           "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED is now ignored.  "
-          "Please set RCUTILS_CONSOLE_USE_STDOUT and RCUTILS_CONSOLE_BUFFERED_STREAM "
+          "Please set RCUTILS_LOGGING_USE_STDOUT and RCUTILS_LOGGING_BUFFERED_STREAM "
           "to control the stream and the buffering of log messages.\n");
       }
     } else {
@@ -168,16 +169,16 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     }
 
     // Set the default output stream for all severities to stderr so that errors
-    // are propagated immediately.  The user can choose to set the output stream
-    // to stdout by setting the RCUTILS_CONSOLE_USE_STDOUT environment
-    // variable to 1.
+    // are propagated immediately.
+    // The user can choose to set the output stream to stdout by setting the
+    // RCUTILS_LOGGING_USE_STDOUT environment variable to 1.
     enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
-      "RCUTILS_CONSOLE_USE_STDOUT", "use stderr",
+      "RCUTILS_LOGGING_USE_STDOUT", "use stderr",
       "use stdout");
     switch (retval) {
       case RCUTILS_GET_ENV_ERROR:
         return RCUTILS_RET_INVALID_ARGUMENT;
-      case RCUTILS_GET_ENV_DEFAULT:
+      case RCUTILS_GET_ENV_EMPTY:
       case RCUTILS_GET_ENV_ZERO:
         g_output_stream = stderr;
         break;
@@ -187,17 +188,18 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     }
 
     // Allow the user to choose how buffering on the stream works by setting
-    // RCUTILS_CONSOLE_BUFFERED_STREAM.  With an empty environment variable, use the
-    // default of the stream.  With a value of 0, force the stream to be unbuffered.
-    // With a value of 1, force the stream to be buffered.
+    // RCUTILS_LOGGING_BUFFERED_STREAM.
+    // With an empty environment variable, use the default of the stream.
+    // With a value of 0, force the stream to be unbuffered.
+    // With a value of 1, force the stream to be line buffered.
     retval =
       rcutils_get_env_var_zero_or_one(
-      "RCUTILS_CONSOLE_BUFFERED_STREAM", "not buffered",
+      "RCUTILS_LOGGING_BUFFERED_STREAM", "not buffered",
       "buffered");
     switch (retval) {
       case RCUTILS_GET_ENV_ERROR:
         return RCUTILS_RET_INVALID_ARGUMENT;
-      case RCUTILS_GET_ENV_DEFAULT:
+      case RCUTILS_GET_ENV_EMPTY:
         break;
       case RCUTILS_GET_ENV_ZERO:
         if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
@@ -229,7 +231,7 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     switch (retval) {
       case RCUTILS_GET_ENV_ERROR:
         return RCUTILS_RET_INVALID_ARGUMENT;
-      case RCUTILS_GET_ENV_DEFAULT:
+      case RCUTILS_GET_ENV_EMPTY:
         g_colorized_output = RCUTILS_COLORIZED_OUTPUT_AUTO;
         break;
       case RCUTILS_GET_ENV_ZERO:

--- a/src/logging.c
+++ b/src/logging.c
@@ -76,14 +76,50 @@ bool g_rcutils_logging_severities_map_valid = false;
 
 int g_rcutils_logging_default_logger_level = 0;
 
-bool g_force_stdout_line_buffered = false;
-bool g_stdout_flush_failure_reported = false;
+static FILE * g_output_stream = NULL;
+static bool g_output_flush_stream = false;
+// A one-time print if we fail to flush
+static bool g_stream_flush_failure_reported = false;
 
 enum rcutils_colorized_output g_colorized_output = RCUTILS_COLORIZED_OUTPUT_AUTO;
 
 rcutils_ret_t rcutils_logging_initialize(void)
 {
   return rcutils_logging_initialize_with_allocator(rcutils_get_default_allocator());
+}
+
+// A utility function to get zero or one from an environment variable.  Returns
+// -1 if we failed to get the environment variable, 0 if the value in the
+// environment variable is something other than "0" or "1", 0 if the environment
+// variable is "0" or "", or "1" if the environment variable is "1".
+static int rcutils_get_env_var_zero_or_one(
+  const char * name, const char * zero_semantic,
+  const char * one_semantic)
+{
+  const char * env_var_value = NULL;
+  const char * ret_str = rcutils_get_env(name, &env_var_value);
+  if (NULL != ret_str) {
+    fprintf(
+      stderr, "Error getting environment variable "
+      "%s: %s\n", name, ret_str);
+    return -1;
+  }
+
+  if (strcmp(env_var_value, "0") == 0 || strcmp(env_var_value, "") == 0) {
+    return 0;
+  }
+
+  if (strcmp(env_var_value, "1") == 0) {
+    return 1;
+  }
+
+  fprintf(
+    stderr,
+    "Warning: unexpected value [%s] specified for %s. "
+    "Valid values are 0 (%s) or 1 (%s).  Default value 0 will be used.\n",
+    env_var_value, name, zero_semantic, one_semantic);
+
+  return 0;
 }
 
 rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allocator)
@@ -99,54 +135,45 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     g_rcutils_logging_output_handler = &rcutils_logging_console_output_handler;
     g_rcutils_logging_default_logger_level = RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL;
 
-    // Check the environment variable for line buffered output
-    const char * line_buffered;
-    const char * ret_str = rcutils_get_env("RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED", &line_buffered);
-
-    if (NULL == ret_str) {
-      if (strcmp(line_buffered, "1") == 0) {
-        g_force_stdout_line_buffered = true;
-      } else if (strcmp(line_buffered, "0") != 0 && strcmp(line_buffered, "") != 0) {
-        fprintf(
-          stderr,
-          "Warning: unexpected value [%s] specified for RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED. "
-          "Default value 0 will be used. Valid values are 1 or 0.\n",
-          line_buffered);
-      }
-    } else {
-      fprintf(
-        stderr, "Error getting env. variable "
-        "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s\n", ret_str);
-      ret = RCUTILS_RET_INVALID_ARGUMENT;
+    // Set the default output stream for all severities to stderr so that errors
+    // are propagated immediately.  The user can choose to set the output stream
+    // to stdout by setting the RCUTILS_CONSOLE_USE_STDOUT environment
+    // variable to 1.
+    g_output_stream = stderr;
+    int retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_CONSOLE_USE_STDOUT", "use stderr",
+      "use stdout");
+    if (retval < 0) {
+      return RCUTILS_RET_INVALID_ARGUMENT;
+    } else if (retval == 0) {
+      g_output_stream = stderr;
+    } else if (retval == 1) {
+      g_output_stream = stdout;
     }
 
-    // Check the environment variable for colorized output
-    const char * colorized_output;
-    ret_str = rcutils_get_env("RCUTILS_COLORIZED_OUTPUT", &colorized_output);
+    retval = rcutils_get_env_var_zero_or_one("RCUTILS_CONSOLE_FLUSH_STREAM", "no flush", "flush");
+    if (retval < 0) {
+      return RCUTILS_RET_INVALID_ARGUMENT;
+    } else if (retval == 0) {
+      g_output_flush_stream = false;
+    } else if (retval == 1) {
+      g_output_flush_stream = true;
+    }
 
-    if (NULL == ret_str) {
-      if (strcmp(colorized_output, "1") == 0) {
-        g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
-      } else if (strcmp(colorized_output, "0") == 0) {
-        g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE;
-      } else if (strcmp(colorized_output, "") != 0) {
-        fprintf(
-          stderr,
-          "Warning: unexpected value [%s] specified for RCUTILS_COLORIZED_OUTPUT. "
-          "Output will be colorized if target stream is a terminal."
-          " Valid values are 0 and 1.\n",
-          colorized_output);
-      }
-    } else {
-      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-        "Failed to get if output is colorized from env. variable [%s]. Using DEFAULT.",
-        ret_str);
-      ret = RCUTILS_RET_INVALID_ARGUMENT;
+    retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_COLORIZED_OUTPUT", "force color",
+      "force no color");
+    if (retval < 0) {
+      return RCUTILS_RET_INVALID_ARGUMENT;
+    } else if (retval == 0) {
+      g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE;
+    } else if (retval == 1) {
+      g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
     }
 
     // Check for the environment variable for custom output formatting
     const char * output_format;
-    ret_str = rcutils_get_env("RCUTILS_CONSOLE_OUTPUT_FORMAT", &output_format);
+    const char * ret_str = rcutils_get_env("RCUTILS_CONSOLE_OUTPUT_FORMAT", &output_format);
     if (NULL == ret_str && strcmp(output_format, "") != 0) {
       size_t chars_to_copy = strlen(output_format);
       if (chars_to_copy > RCUTILS_LOGGING_MAX_OUTPUT_FORMAT_LEN - 1) {
@@ -687,14 +714,14 @@ rcutils_ret_t rcutils_logging_format_message(
 # define IS_STREAM_A_TTY(stream) (isatty(fileno(stream)) != 0)
 #endif
 
-#define IS_OUTPUT_COLORIZED(is_colorized, stream) \
+#define IS_OUTPUT_COLORIZED(is_colorized) \
   { \
     if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE) { \
       is_colorized = true; \
     } else if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE) { \
       is_colorized = false; \
     } else { \
-      is_colorized = IS_STREAM_A_TTY(stream); \
+      is_colorized = IS_STREAM_A_TTY(g_output_stream); \
     } \
   }
 #define SET_COLOR_WITH_SEVERITY(status, severity, color) \
@@ -729,10 +756,10 @@ rcutils_ret_t rcutils_logging_format_message(
       } \
     } \
   }
-# define GET_HANDLE_FROM_STREAM(status, handle, stream) \
+# define GET_HANDLE_FROM_STREAM(status, handle) \
   { \
     if (RCUTILS_RET_OK == status) { \
-      if (stream == stdout) { \
+      if (g_output_stream == stdout) { \
         handle = GetStdHandle(STD_OUTPUT_HANDLE); \
       } else { \
         handle = GetStdHandle(STD_ERROR_HANDLE); \
@@ -744,19 +771,19 @@ rcutils_ret_t rcutils_logging_format_message(
       } \
     } \
   }
-# define SET_OUTPUT_COLOR_WITH_SEVERITY(status, severity, stream, output_array) \
+# define SET_OUTPUT_COLOR_WITH_SEVERITY(status, severity, output_array) \
   { \
     WORD color = COLOR_NORMAL; \
     HANDLE handle = INVALID_HANDLE_VALUE; \
     SET_COLOR_WITH_SEVERITY(status, severity, color) \
-    GET_HANDLE_FROM_STREAM(status, handle, stream) \
+    GET_HANDLE_FROM_STREAM(status, handle) \
     SET_OUTPUT_COLOR_WITH_COLOR(status, color, handle) \
   }
-# define SET_STANDARD_COLOR_IN_STREAM(is_colorized, status, stream) \
+# define SET_STANDARD_COLOR_IN_STREAM(is_colorized, status) \
   { \
     if (is_colorized) { \
       HANDLE handle = INVALID_HANDLE_VALUE; \
-      GET_HANDLE_FROM_STREAM(status, handle, stream) \
+      GET_HANDLE_FROM_STREAM(status, handle) \
       SET_OUTPUT_COLOR_WITH_COLOR(status, COLOR_NORMAL, handle) \
     } \
   }
@@ -773,7 +800,7 @@ rcutils_ret_t rcutils_logging_format_message(
       } \
     } \
   }
-# define SET_OUTPUT_COLOR_WITH_SEVERITY(status, severity, stream, output_array) \
+# define SET_OUTPUT_COLOR_WITH_SEVERITY(status, severity, output_array) \
   { \
     const char * color = NULL; \
     SET_COLOR_WITH_SEVERITY(status, severity, color) \
@@ -785,7 +812,7 @@ rcutils_ret_t rcutils_logging_format_message(
       SET_OUTPUT_COLOR_WITH_COLOR(status, COLOR_NORMAL, output_array) \
     } \
   }
-# define SET_STANDARD_COLOR_IN_STREAM(is_colorized, status, stream)
+# define SET_STANDARD_COLOR_IN_STREAM(is_colorized, status)
 #endif
 
 void rcutils_logging_console_output_handler(
@@ -803,23 +830,18 @@ void rcutils_logging_console_output_handler(
       "call to rcutils_logging_console_output_handler failed.\n");
     return;
   }
-  FILE * stream = NULL;
   switch (severity) {
     case RCUTILS_LOG_SEVERITY_DEBUG:
     case RCUTILS_LOG_SEVERITY_INFO:
-      stream = stdout;
-      break;
     case RCUTILS_LOG_SEVERITY_WARN:
     case RCUTILS_LOG_SEVERITY_ERROR:
     case RCUTILS_LOG_SEVERITY_FATAL:
-      stream = stderr;
       break;
     default:
       fprintf(stderr, "unknown severity level: %d\n", severity);
-      return;
   }
 
-  IS_OUTPUT_COLORIZED(is_colorized, stream)
+  IS_OUTPUT_COLORIZED(is_colorized)
 
   char msg_buf[1024] = "";
   rcutils_char_array_t msg_array = {
@@ -840,7 +862,7 @@ void rcutils_logging_console_output_handler(
   };
 
   if (is_colorized) {
-    SET_OUTPUT_COLOR_WITH_SEVERITY(status, severity, stream, output_array)
+    SET_OUTPUT_COLOR_WITH_SEVERITY(status, severity, output_array)
   }
 
   if (RCUTILS_RET_OK == status) {
@@ -869,12 +891,12 @@ void rcutils_logging_console_output_handler(
   SET_STANDARD_COLOR_IN_BUFFER(is_colorized, status, output_array)
 
   if (RCUTILS_RET_OK == status) {
-    fprintf(stream, "%s\n", output_array.buffer);
+    fprintf(g_output_stream, "%s\n", output_array.buffer);
 
-    if (g_force_stdout_line_buffered && stream == stdout) {
-      int flush_result = fflush(stream);
-      if (flush_result != 0 && !g_stdout_flush_failure_reported) {
-        g_stdout_flush_failure_reported = true;
+    if (g_output_flush_stream) {
+      int flush_result = fflush(g_output_stream);
+      if (flush_result != 0 && !g_stream_flush_failure_reported) {
+        g_stream_flush_failure_reported = true;
         fprintf(
           stderr, "Error: failed to perform fflush on stdout, fflush return code is: %d\n",
           flush_result);
@@ -884,7 +906,7 @@ void rcutils_logging_console_output_handler(
 
   // Only does something in windows
   // cppcheck-suppress uninitvar  // suppress cppcheck false positive
-  SET_STANDARD_COLOR_IN_STREAM(is_colorized, status, stream)
+  SET_STANDARD_COLOR_IN_STREAM(is_colorized, status)
 
   status = rcutils_char_array_fini(&msg_array);
   if (RCUTILS_RET_OK != status) {

--- a/src/logging.c
+++ b/src/logging.c
@@ -206,7 +206,7 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
       int mode = retval == RCUTILS_GET_ENV_ZERO ? _IONBF : _IOLBF;
       if (setvbuf(g_output_stream, NULL, mode, 0) != 0) {
         char error_string[1024];
-        rcutils_safe_strerror(error_string, sizeof(error_string));
+        rcutils_strerror(error_string, sizeof(error_string));
         fprintf(
           stderr, "Error setting stream buffering mode: %s\n", error_string);
         RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(

--- a/src/logging.c
+++ b/src/logging.c
@@ -39,6 +39,7 @@ extern "C"
 #include "rcutils/logging.h"
 #include "rcutils/snprintf.h"
 #include "rcutils/strdup.h"
+#include "rcutils/strerror.h"
 #include "rcutils/time.h"
 #include "rcutils/types/string_map.h"
 
@@ -200,21 +201,23 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
         break;
       case RCUTILS_GET_ENV_ZERO:
         if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
+          char error_string[1024];
+          rcutils_safe_strerror(error_string, sizeof(error_string));
           fprintf(
-            stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+            stderr, "Error setting stream buffering mode: %s\n", error_string);
           RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Error setting stream buffering mode: %s", strerror(
-              errno));
+            "Error setting stream buffering mode: %s", error_string);
           return RCUTILS_RET_ERROR;
         }
         break;
       case RCUTILS_GET_ENV_ONE:
         if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
+          char error_string[1024];
+          rcutils_safe_strerror(error_string, sizeof(error_string));
           fprintf(
-            stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+            stderr, "Error setting stream buffering mode: %s\n", error_string);
           RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Error setting stream buffering mode: %s", strerror(
-              errno));
+            "Error setting stream buffering mode: %s", error_string);
           return RCUTILS_RET_ERROR;
         }
         break;

--- a/src/logging.c
+++ b/src/logging.c
@@ -173,12 +173,16 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
       "RCUTILS_CONSOLE_USE_STDOUT", "use stderr",
       "use stdout");
-    if (retval == RCUTILS_GET_ENV_ERROR) {
-      return RCUTILS_RET_INVALID_ARGUMENT;
-    } else if (retval == RCUTILS_GET_ENV_DEFAULT || retval == RCUTILS_GET_ENV_ZERO) {
-      g_output_stream = stderr;
-    } else if (retval == RCUTILS_GET_ENV_ONE) {
-      g_output_stream = stdout;
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_DEFAULT:
+      case RCUTILS_GET_ENV_ZERO:
+        g_output_stream = stderr;
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        g_output_stream = stdout;
+        break;
     }
 
     // Allow the user to choose how buffering on the stream works by setting
@@ -189,40 +193,49 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
       rcutils_get_env_var_zero_or_one(
       "RCUTILS_CONSOLE_BUFFERED_STREAM", "not buffered",
       "buffered");
-    if (retval == RCUTILS_GET_ENV_ERROR) {
-      return RCUTILS_RET_INVALID_ARGUMENT;
-    } else if (retval == RCUTILS_GET_ENV_ZERO) {
-      if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
-        fprintf(
-          stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
-        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-          "Error setting stream buffering mode: %s", strerror(
-            errno));
-        return RCUTILS_RET_ERROR;
-      }
-    } else if (retval == RCUTILS_GET_ENV_ONE) {
-      if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
-        fprintf(
-          stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
-        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-          "Error setting stream buffering mode: %s", strerror(
-            errno));
-        return RCUTILS_RET_ERROR;
-      }
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_DEFAULT:
+        break;
+      case RCUTILS_GET_ENV_ZERO:
+        if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
+          fprintf(
+            stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "Error setting stream buffering mode: %s", strerror(
+              errno));
+          return RCUTILS_RET_ERROR;
+        }
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
+          fprintf(
+            stderr, "Error setting stream buffering mode: %s\n", strerror(errno));
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "Error setting stream buffering mode: %s", strerror(
+              errno));
+          return RCUTILS_RET_ERROR;
+        }
+        break;
     }
-    // Note that for retval == RCUTILS_GET_ENV_DEFAULT, we do nothing to leave it at default.
 
     retval = rcutils_get_env_var_zero_or_one(
       "RCUTILS_COLORIZED_OUTPUT", "force color",
       "force no color");
-    if (retval == RCUTILS_GET_ENV_ERROR) {
-      return RCUTILS_RET_INVALID_ARGUMENT;
-    } else if (retval == RCUTILS_GET_ENV_ZERO) {
-      g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE;
-    } else if (retval == RCUTILS_GET_ENV_ONE) {
-      g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_DEFAULT:
+        g_colorized_output = RCUTILS_COLORIZED_OUTPUT_AUTO;
+        break;
+      case RCUTILS_GET_ENV_ZERO:
+        g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE;
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
+        break;
     }
-    // Note that for retval == RCUTILS_GET_ENV_DEFAULT, we do nothing to leave it at default.
 
     // Check for the environment variable for custom output formatting
     const char * output_format;
@@ -892,6 +905,7 @@ void rcutils_logging_console_output_handler(
       break;
     default:
       fprintf(stderr, "unknown severity level: %d\n", severity);
+      return;
   }
 
   IS_OUTPUT_COLORIZED(is_colorized)

--- a/src/logging.c
+++ b/src/logging.c
@@ -109,7 +109,9 @@ static enum rcutils_get_env_retval rcutils_get_env_var_zero_or_one(
     fprintf(
       stderr, "Error getting environment variable "
       "%s: %s\n", name, ret_str);
-    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("Error getting environment variable: %s", ret_str);
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting environment variable %s: %s", name,
+      ret_str);
     return RCUTILS_GET_ENV_ERROR;
   }
 
@@ -144,6 +146,25 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
 
     g_rcutils_logging_output_handler = &rcutils_logging_console_output_handler;
     g_rcutils_logging_default_logger_level = RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL;
+
+    const char * line_buffered = NULL;
+    const char * ret_str = rcutils_get_env("RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED", &line_buffered);
+    if (NULL == ret_str) {
+      if (strcmp(line_buffered, "") != 0) {
+        fprintf(
+          stderr,
+          "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED is now ignored.  "
+          "Please set RCUTILS_CONSOLE_USE_STDOUT and RCUTILS_CONSOLE_BUFFERED_STREAM "
+          "to control the stream and the buffering of log messages.\n");
+      }
+    } else {
+      fprintf(
+        stderr, "Error getting environment variable "
+        "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s\n", ret_str);
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Error getting environment variable RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s", ret_str);
+      return RCUTILS_RET_ERROR;
+    }
 
     // Set the default output stream for all severities to stderr so that errors
     // are propagated immediately.  The user can choose to set the output stream
@@ -205,7 +226,7 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
 
     // Check for the environment variable for custom output formatting
     const char * output_format;
-    const char * ret_str = rcutils_get_env("RCUTILS_CONSOLE_OUTPUT_FORMAT", &output_format);
+    ret_str = rcutils_get_env("RCUTILS_CONSOLE_OUTPUT_FORMAT", &output_format);
     if (NULL == ret_str && strcmp(output_format, "") != 0) {
       size_t chars_to_copy = strlen(output_format);
       if (chars_to_copy > RCUTILS_LOGGING_MAX_OUTPUT_FORMAT_LEN - 1) {

--- a/src/logging.c
+++ b/src/logging.c
@@ -173,8 +173,7 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     // The user can choose to set the output stream to stdout by setting the
     // RCUTILS_LOGGING_USE_STDOUT environment variable to 1.
     enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
-      "RCUTILS_LOGGING_USE_STDOUT", "use stderr",
-      "use stdout");
+      "RCUTILS_LOGGING_USE_STDOUT", "use stderr", "use stdout");
     switch (retval) {
       case RCUTILS_GET_ENV_ERROR:
         return RCUTILS_RET_INVALID_ARGUMENT;
@@ -185,6 +184,12 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
       case RCUTILS_GET_ENV_ONE:
         g_output_stream = stdout;
         break;
+      default:
+        fprintf(
+          stderr, "Invalid return from environment fetch\n");
+        RCUTILS_SET_ERROR_MSG(
+          "Invalid return from environment fetch");
+        return RCUTILS_RET_ERROR;
     }
 
     // Allow the user to choose how buffering on the stream works by setting
@@ -192,37 +197,28 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     // With an empty environment variable, use the default of the stream.
     // With a value of 0, force the stream to be unbuffered.
     // With a value of 1, force the stream to be line buffered.
-    retval =
-      rcutils_get_env_var_zero_or_one(
-      "RCUTILS_LOGGING_BUFFERED_STREAM", "not buffered",
-      "buffered");
-    switch (retval) {
-      case RCUTILS_GET_ENV_ERROR:
-        return RCUTILS_RET_INVALID_ARGUMENT;
-      case RCUTILS_GET_ENV_EMPTY:
-        break;
-      case RCUTILS_GET_ENV_ZERO:
-        if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
-          char error_string[1024];
-          rcutils_safe_strerror(error_string, sizeof(error_string));
-          fprintf(
-            stderr, "Error setting stream buffering mode: %s\n", error_string);
-          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Error setting stream buffering mode: %s", error_string);
-          return RCUTILS_RET_ERROR;
-        }
-        break;
-      case RCUTILS_GET_ENV_ONE:
-        if (setvbuf(g_output_stream, NULL, _IOLBF, 0) != 0) {
-          char error_string[1024];
-          rcutils_safe_strerror(error_string, sizeof(error_string));
-          fprintf(
-            stderr, "Error setting stream buffering mode: %s\n", error_string);
-          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
-            "Error setting stream buffering mode: %s", error_string);
-          return RCUTILS_RET_ERROR;
-        }
-        break;
+    retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_LOGGING_BUFFERED_STREAM", "not buffered", "buffered");
+    if (RCUTILS_GET_ENV_ERROR == retval) {
+      return RCUTILS_RET_INVALID_ARGUMENT;
+    }
+    if (RCUTILS_GET_ENV_ZERO == retval || RCUTILS_GET_ENV_ONE == retval) {
+      int mode = retval == RCUTILS_GET_ENV_ZERO ? _IONBF : _IOLBF;
+      if (setvbuf(g_output_stream, NULL, mode, 0) != 0) {
+        char error_string[1024];
+        rcutils_safe_strerror(error_string, sizeof(error_string));
+        fprintf(
+          stderr, "Error setting stream buffering mode: %s\n", error_string);
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Error setting stream buffering mode: %s", error_string);
+        return RCUTILS_RET_ERROR;
+      }
+    } else if (RCUTILS_GET_ENV_EMPTY != retval) {
+      fprintf(
+        stderr, "Invalid return from environment fetch\n");
+      RCUTILS_SET_ERROR_MSG(
+        "Invalid return from environment fetch");
+      return RCUTILS_RET_ERROR;
     }
 
     retval = rcutils_get_env_var_zero_or_one(
@@ -240,6 +236,12 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
       case RCUTILS_GET_ENV_ONE:
         g_colorized_output = RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE;
         break;
+      default:
+        fprintf(
+          stderr, "Invalid return from environment fetch\n");
+        RCUTILS_SET_ERROR_MSG(
+          "Invalid return from environment fetch");
+        return RCUTILS_RET_ERROR;
     }
 
     // Check for the environment variable for custom output formatting

--- a/src/strerror.c
+++ b/src/strerror.c
@@ -12,13 +12,15 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include <errno.h>
-#include <string.h>
-
 #ifdef __cplusplus
 extern "C"
 {
 #endif
+
+#include <errno.h>
+#include <string.h>
+
+#include "rcutils/strerror.h"
 
 void
 rcutils_safe_strerror(char * buffer, size_t buffer_length)

--- a/src/strerror.c
+++ b/src/strerror.c
@@ -1,0 +1,47 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+void
+rcutils_safe_strerror(char * buffer, size_t buffer_length)
+{
+#if defined(_WIN32)
+  strerror_s(buffer, buffer_length, errno);
+#elif defined(_GNU_SOURCE) && (!defined(ANDROID) || __ANDROID_API__ >= 23)
+  /* GNU-specific */
+  char * msg = strerror_r(errno, buffer, buffer_length);
+  if (msg != buffer) {
+    strncpy(buffer, msg, buffer_length);
+    buffer[buffer_length - 1] = '\0';
+  }
+#else
+  /* XSI-compliant */
+  int error_status = strerror_r(errno, buffer, buffer_length);
+  if (error_status != 0) {
+    strncpy(buffer, "Failed to get error", buffer_length);
+    buffer[buffer_length - 1] = '\0';
+  }
+#endif
+}
+
+#ifdef __cplusplus
+}
+#endif

--- a/src/strerror.c
+++ b/src/strerror.c
@@ -23,7 +23,7 @@ extern "C"
 #include "rcutils/strerror.h"
 
 void
-rcutils_safe_strerror(char * buffer, size_t buffer_length)
+rcutils_strerror(char * buffer, size_t buffer_length)
 {
 #if defined(_WIN32)
   strerror_s(buffer, buffer_length, errno);

--- a/test/test_logging_output_format.py
+++ b/test/test_logging_output_format.py
@@ -92,7 +92,7 @@ class TestLoggingOutputFormatAfterShutdown(unittest.TestCase):
     def test_logging_output(self, proc_output, processes_to_test):
         """Test all executables output against expectations."""
         for process_name in processes_to_test:
-            launch_testing.asserts.assertInStdout(
+            launch_testing.asserts.assertInStderr(
                 proc_output,
                 expected_output=launch_testing.tools.expected_output_from_file(
                     path=os.path.join(os.path.dirname(__file__), process_name)

